### PR TITLE
[Benchmark] Add encoder decoder to benchmark and clean labels

### PR DESCRIPTION
**Pull request by patrickvonplaten**
This PR cleans the benchmark utils a bit more:
- tracing is made independent from CPU memory benchmarking
- possibility to benchmark encoder-decoder models is added
- 3 new tests
- general refactoring

### CODE REVIEW

File: src/transformers/benchmark/benchmark.py [hunk 1]
```diff
@@ -52,46 +52,79 @@ def train(self, model_name, batch_size, sequence_length, trace_memory=False):
...
 
+            # encoder-decoder has vocab size saved differently
+            vocab_size = config.vocab_size if hasattr(config, "vocab_size") else config.encoder.vocab_size
             input_ids = torch.randint(
-                model.config.vocab_size, (batch_size, sequence_length), dtype=torch.long, device=self.args.device
+                vocab_size, (batch_size, sequence_length), dtype=torch.long, device=self.args.device
             )
 
-            def compute_loss_and_backprob():
-                # TODO: Not all models call labels argument labels => this hack using the function signature should be corrected once all models have a common name for labels
```
Comment (patrickvonplaten) [inline src/transformers/benchmark/benchmark.py:60]: Thanks to @sgugger this function inspect logit is not necessary anymore.

File: src/transformers/benchmark/benchmark.py [hunk 2]
```diff
@@ -52,46 +52,79 @@ def train(self, model_name, batch_size, sequence_length, trace_memory=False):
...
 
+            _train = (
+                compute_loss_and_backprob_encoder_decoder
+                if config.is_encoder_decoder
+                else compute_loss_and_backprob_encoder
+            )
+
             if trace_memory is True:
-                if self.args.trace_memory_line_by_line or self.args.n_gpu == 0:
+                if self.args.trace_memory_line_by_line:
```
Comment (patrickvonplaten) [inline src/transformers/benchmark/benchmark.py:78]: tracing should be complementary and not always be used for cpu

File: src/transformers/benchmark/benchmark_args_utils.py [hunk 3]
```diff
@@ -61,6 +61,12 @@ class BenchmarkArguments:
     save_to_csv: bool = field(default=False, metadata={"help": "Save result to a CSV file"})
     log_print: bool = field(default=False, metadata={"help": "Save all print statements in a log file"})
     no_env_print: bool = field(default=False, metadata={"help": "Don't print environment information"})
+    with_lm_head: bool = field(
```
Comment (patrickvonplaten) [inline src/transformers/benchmark/benchmark_args_utils.py:64]: allow the possibility to switch between model + lm_head and without

File: tests/test_benchmark.py [hunk 4]
```diff
@@ -88,3 +110,29 @@ def test_save_csv_files(self):
             self.assertTrue(Path(os.path.join(tmp_dir, "inf_mem.csv")).exists())
             self.assertTrue(Path(os.path.join(tmp_dir, "train_mem.csv")).exists())
             self.assertTrue(Path(os.path.join(tmp_dir, "env.csv")).exists())
+
+    def test_trace_memory(self):
```
Comment (patrickvonplaten) [inline tests/test_benchmark.py:114]: add tracing test

File: src/transformers/benchmark/benchmark_args_utils.py (hunk 3 above)
Comment (LysandreJik) [reply src/transformers/benchmark/benchmark_args_utils.py:64]: Cool, I think this makes sense.

**Review by LysandreJik — APPROVED** [review]
Cool, that's clean! I like the improvement.

**Pull request merged**